### PR TITLE
bazel: use minimal C toolchains and make macOS builds fully hermetic

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR updates a few things about our Bazel setup, namely the C toolchain:

* Provide the macOS 14.5 SDK as the sysroot and set the minimum macOS version we build for as `14.5`. This should make macOS builds fully hermetic and maximize remote cache hits between developer machines.
* Use a minimal `clang 18.1.8` toolchain. Previously we used upstream clang 16 which was ~800MB compressed with `xz`, now the toolchain is ~70MB and compressed with `zstd`. This change should save us 1.5 - 2 minutes in Bazel startup time.

Both the sysroot and clang toolchain are built with GitHub actions from the [MaterializeInc/toolchains](https://github.com/MaterializeInc/toolchains) repo.

This PR also updates the version of Bazel we use which generates a smaller `MODULE.bazel.lock`

Note: Ideally we also provide a sysroot for Linux builds, I just haven't gotten around to building that though.

### Motivation

Faster build times, better remote caching, and possibly faster runtime since we're using a newer version of `clang`.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

**Comment by ParkMyCar** [comment]
Need to test that this actually improves the Bazel remote caching before merging, and move the sysroots to a Materialize owned repository.

**Comment by ParkMyCar** [comment]
Tested the sysroot approach with @danhhz a little while ago and it made remote caching nearly perfect for developer builds!

File: WORKSPACE [hunk 1]
```diff
@@ -164,6 +188,20 @@ load("@toolchains_llvm//toolchain:rules.bzl", "llvm_toolchain")
...
+    },
+    urls = {
+        "darwin-aarch64": ["https://github.com/MaterializeInc/toolchains/releases/download/clang-{0}/darwin_aarch64.tar.zst".format(LLVM_VERSION)],
+        "linux-aarch64": ["https://github.com/MaterializeInc/toolchains/releases/download/clang-{0}/linux_aarch64.tar.zst".format(LLVM_VERSION)],
+        "linux-x86_64": ["https://github.com/MaterializeInc/toolchains/releases/download/clang-{0}/linux_x86_64.tar.zst".format(LLVM_VERSION)],
+    },
+    sha256 = {
+        "darwin-aarch64": "c538f0a8e359f6fec68f4fb6a7e1a36cbb1e7fd1d74640d6b413c204c599d4f4",
+        "linux-aarch64": "e899e3e40748d4c9e704316f0e14b565b2646a0d2d2525a5d9cbe3dcf5389842",
+        "linux-x86_64": "",
```
Comment (danhhz) [inline WORKSPACE:203]: intentionally missing?

**Review by danhhz — APPROVED** [review]
I'm pretty out of my depth in this build stuff and I only skimmed the toolchains repo, but stamping this to unblock bazel progress.

Might be worth an asynchronous @benesch skim at some point (mostly for the toolchains repo), but I wouldn't block merging on that.

**Review by benesch — APPROVED** [review]
I've been lightly following the toolchains repo and it all makes sense to me!

File: WORKSPACE (hunk 1 above)
Comment (ParkMyCar) [reply WORKSPACE]: Whoops, definitely not intentional, thanks for catching this!

**Comment by ParkMyCar** [comment]
Thanks for the reviews!

**Pull request merged**